### PR TITLE
Feat(eos_designs): Add option for ip igmp snooping fast-leave

### LINKED ISSUE #3942: Feat(eos_designs): add option for ip igmp snooping fast-leave
### Enhancement summary

Add key to enable ip igmp snooping vlan <x> fast-leave when 
```
evpn_l2_multicast: 
  enabled:  true
  ```
  is configured

### Which component of AVD is impacted

eos_designs

### Use case example

When deploying l2 multicast, configuration stanzas are issued for:
```
ip igmp snooping vlan 1234 querier
ip igmp snooping vlan 1234 querier address 1.2.3.4
```
There is a need to add `ip igmp snooping vlan 1234 querier` 

### Describe the solution you would like

Add the key `fast-leave` to the existing tenant configuration:

```
            igmp_snooping_querier:
              # Will be enabled automatically if "evpn_l2_multicast" is enabled.
              enabled: < true | false | default false >
              source_address: < ipv4_address -> default ip address of Loopback0 >
              version: < 1, 2, 3 -> default 2 (EOS) >
```

### Describe alternatives you have considered

We can generate the configuration with:

```
            eos_cli: |
              ip igmp snooping vlan 123 fast-leave
              ip igmp snooping vlan 1234 querier version 3
              ip igmp snooping vlan 1234 fast-leave
```

### Additional context

_No response_

### Contributing Guide

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Shivani-gslab**
## Change Summary

Add option for ip igmp snooping fast-leave

## Related Issue(s)

Fixes #3942 

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
Adding `fast_leave` key to `evnpn_l2_multicast` and inner `igmp_snooping_querier` in tenants (network_services).
Inserting `fast_leave` config only if `evpn_l2_multicast.enabled` or inner `igmp_snooping_querier.enabled is true` and `fast_leave is true`.

## How to test
Added tests in `eos_designs_unit_tests` molecule scenario in this PR. Test with EOS CLI.

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

**Review by Vibhu-gslab — APPROVED** [review]

File: python-avd/pyavd/_eos_designs/structured_config/network_services/ip_igmp_snooping.py [hunk 1]
```diff
@@ -114,6 +114,14 @@ def _ip_igmp_snooping_vlan(self: AvdStructuredConfigNetworkServices, vlan, tenan
             if version is not None:
                 ip_igmp_snooping_vlan["querier"]["version"] = version
 
+            fast_leave = default(
+                igmp_snooping_querier.get("fast_leave"),
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/ip_igmp_snooping.py:120]: Move this condition out of the if condition on line 102. As condition on line 102 can be true even when evpn_l2_multicast is not enabled.

Suggestion:

if evpn_l2_multicast_enabled:
            fast_leave = default(
                igmp_snooping_querier.get("fast_leave"),
                tenant_igmp_snooping_querier.get("fast_leave"),
                get(tenant, "evpn_l2_multicast.fast_leave"),
            )
            if fast_leave is not None:
                ip_igmp_snooping_vlan["fast_leave"] = fast_leave

Comment (Shivani-gslab) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/ip_igmp_snooping.py:120]: fixed!

**Comment by laxmikantchintakindi** [comment]
LGTM!

**Review by laxmikantchintakindi — APPROVED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/EVPN-MULTICAST-L3LEAF1B.cfg [hunk 2]
```diff
@@ -200,7 +220,7 @@ vlan 4093
    trunk group LEAF_PEER_L3
 !
 vlan 4094
-   name MLAG_PEER
```
Comment (laxmikantchintakindi) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/EVPN-MULTICAST-L3LEAF1B.cfg]: Use different interface. As it is replacing the old one.

Comment (Shivani-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/EVPN-MULTICAST-L3LEAF1B.cfg]: fixed

**Review by laxmikantchintakindi — APPROVED** [review]

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 3]
```diff
@@ -172,6 +172,9 @@ $defs:
               type: int
               convert_types:
                 - str
+            fast_leave:
+              type: bool
+              description: Enable fast-leave processing on specified VLANs.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]:
```suggestion
              description: Enable IGMP snooping fast-leave feature on specified VLANs.
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 4]
```diff
@@ -1249,6 +1252,9 @@ $defs:
                       - 2
                       - 3
                     default: 2
+                  fast_leave:
+                    type: bool
+                    description: Enable fast-leave processing on specified VLANs.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: same description

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml [hunk 5]
```diff
@@ -202,6 +202,9 @@ $defs:
               - 1
               - 2
               - 3
+          fast_leave:
+            type: bool
+            description: Enable fast-leave processing on specified VLANs.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml]: same description

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml [hunk 6]
```diff
@@ -582,3 +591,23 @@ tenants:
...
+        name: "MULTICAST_DISABLED_251"
+        tags: ['test_l2']
+        evpn_l2_multicast:
+          enabled: false
+      - id: 252
+        name: "MULTICAST_ENABLED_252"
+        tags: ['test_l2']
+      - id: 4091
+        name: "MULTICAST_ENABLED_4091"
+        tags: ['test_l2']
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml]: do we need this one in addition to 252? what does it bring?

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml [hunk 7]
```diff
@@ -93,6 +94,7 @@ tenants:
       - id: 256
         name: "MULTICAST_DISABLED_256"
         tags: ['test_l2']
+        # test for fast-leave when tenant level evpn_l2_multicast.enabled: true, vpn_l2_multicast.fast_leave: true and l2vlans.evpn_l2_multicast.enabled: false
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml]:
```suggestion
        # test for fast-leave when tenant level evpn_l2_multicast.enabled: true, evpn_l2_multicast.fast_leave: true and l2vlans.evpn_l2_multicast.enabled: false
```

Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml]: add the expectation

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml [hunk 8]
```diff
@@ -132,8 +134,11 @@ tenants:
             tags: ['test_l3']
             enabled: true
             ip_address_virtual: 10.0.1.1/24
+            # test for fast-leave when tenant level evpn_l2_multicast.enabled: false and vrfs.svis.evpn_l2_multicast.enabled: true with igmp_snooping_querier.fast_leave: true
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:139]: edit the expectation line 124 to state we expect fast-leave on VLAN 1

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml [hunk 9]
```diff
@@ -132,8 +134,11 @@ tenants:
...
             evpn_l2_multicast:
               enabled: true
+            igmp_snooping_querier:
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:142]: why is enabled not needed here but needed for the l2vlans below?

File: python-avd/pyavd/_eos_designs/structured_config/network_services/ip_igmp_snooping.py [hunk 10]
```diff
@@ -114,6 +114,14 @@ def _ip_igmp_snooping_vlan(self: AvdStructuredConfigNetworkServices, vlan, tenan
             if version is not None:
                 ip_igmp_snooping_vlan["querier"]["version"] = version
 
+        if evpn_l2_multicast_enabled is True:
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/structured_config/network_services/ip_igmp_snooping.py:118]: put some comment in the code explaining we consider fast-leave only if evpn_l2_multicast is enabled

**Review by gmuloc — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml (hunk 9 above)
Comment (Shivani-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:142]: it is not required in l2vlans too

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml (hunk 6 above)
Comment (Shivani-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml]: umm..yeah we can remove it!

Comment (Shivani-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml]: Fixed

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml (hunk 5 above)
Comment (Shivani-gslab) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml]: done

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml (hunk 7 above)
Comment (Shivani-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml]: added

File: python-avd/pyavd/_eos_designs/structured_config/network_services/ip_igmp_snooping.py (hunk 10 above)
Comment (Shivani-gslab) [reply python-avd/pyavd/_eos_designs/structured_config/network_services/ip_igmp_snooping.py:118]: added

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml (hunk 8 above)
Comment (Shivani-gslab) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:139]: fixed

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml (hunk 4 above)
Comment (Shivani-gslab) [reply python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: fixed

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml (hunk 9 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:142]: you have it line 304 right?

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 11]
```diff
@@ -1249,6 +1252,9 @@ $defs:
                       - 2
                       - 3
                     default: 2
+                  fast_leave:
+                    type: bool
+                    description: Enable IGMP snooping fast-leave feature on specified VLANs.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]: Actually sorry for missing this but here this is only for one given VLAN not VLANs
```suggestion
                    description: Enable IGMP snooping fast-leave feature.
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml [hunk 12]
```diff
@@ -172,6 +172,9 @@ $defs:
               type: int
               convert_types:
                 - str
+            fast_leave:
+              type: bool
+              description: Enable IGMP snooping fast-leave feature on specified VLANs.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_network_services.schema.yml]:
```suggestion
              description: Enable IGMP snooping fast-leave feature for all SVIs and l2vlans within Tenant.
```

File: python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml [hunk 13]
```diff
@@ -202,6 +202,9 @@ $defs:
               - 1
               - 2
               - 3
+          fast_leave:
+            type: bool
+            description: Enable IGMP snooping fast-leave feature on specified VLANs.
```
Comment (gmuloc) [inline python-avd/pyavd/_eos_designs/schema/schema_fragments/defs_svi_settings.schema.yml]:
```suggestion
            description: Enable IGMP snooping fast-leave feature.
```

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml (hunk 9 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/EVPN_MULTICAST_TESTS.yml:142]: ok I have read the schema more carefully - when evpn_l2_multicast is true then this is enabled by default. Sorry for the noise

**Review by gmuloc — COMMENTED** [review]
some description schema changed I overlooked in my previous review - last changes and we can merge as far as I am concerned

**Review by gmuloc — APPROVED** [review]

**Pull request merged**